WooCommerce/Classes/ViewRelated/Dashboard/Settings/Plugins/PluginListViewModel.swift

```diff
@@ -1,13 +1,120 @@
+import protocol Storage.StorageManagerType
 
 final class PluginListViewModel {
 
+    /// ID of the site to load plugins for
+    ///
+    private let siteID: Int64
+
     /// Reference to the StoresManager to dispatch Yosemite Actions.
     ///
     private let storesManager: StoresManager
 
-    init(storesManager: StoresManager = ServiceLocator.stores) {
+    /// StorageManager to load plugins from storage
+    ///
+    private let storageManager: StorageManagerType
+
+    /// Results controller for the plugin list
+    ///
+    private lazy var resultsController: ResultsController<StorageSitePlugin> = {
+        let predicate = NSPredicate(format: "siteID = %ld", self.siteID)
+        let nameDescriptor = NSSortDescriptor(keyPath: \StorageSitePlugin.name, ascending: true)
+        // Results need to be grouped in sections so sorting by section is required.
+        // Make sure this sort descriptor is first in the list to make grouping works.
+        let statusDescriptor = NSSortDescriptor(keyPath: \StorageSitePlugin.status, ascending: true)
+        return ResultsController<StorageSitePlugin>(
+            storageManager: storageManager,
+            sectionNameKeyPath: "status",
+            matching: predicate,
+            sortedBy: [statusDescriptor, nameDescriptor]
+        )
+    }()
+
+    init(siteID: Int64,
+         storesManager: StoresManager = ServiceLocator.stores,
+         storageManager: StorageManagerType = ServiceLocator.storageManager) {
+        self.siteID = siteID
         self.storesManager = storesManager
+        self.storageManager = storageManager
+    }
+
+    /// Start fetching and observing plugin data from local storage.
+    ///
+    func observePlugins(onDataChanged: @escaping () -> Void) {
+        do {
+            try resultsController.performFetch()
+        } catch {
+            DDLogError("⛔️ Error fetching plugin list!")
+        }
+
+        resultsController.onDidChangeContent = onDataChanged
     }
 }
+
+// MARK: - Data source for plugin list
+//
+extension PluginListViewModel {
+    /// Number of sections to display on the table view.
+    ///
+    var numberOfSections: Int {
+        resultsController.sections.count
+    }
+
+    /// Title of table view section at specified index.
+    ///
+    func titleForSection(at index: Int) -> String? {
+        guard let rawStatus = resultsController.sections[safe: index]?.name else {
+            return nil
+        }
+        let pluginStatus = SitePluginStatusEnum(rawValue: rawStatus)
+        let sectionTitle: String
+        switch pluginStatus {
+        case .active:
+            sectionTitle = Localization.activeSectionTitle
+        case .inactive:
+            sectionTitle = Localization.inactiveSectionTitle
+        case .networkActive:
+            sectionTitle = Localization.networkActiveSectionTitle
+        case .unknown:
+            sectionTitle = "" // This case should not happen
+        }
+        return sectionTitle.capitalized
```

Comment:
nit: This code is just for figuring out the section title but If I'm reading the code I would have to inspect any single case to know for sure that there is not extra logic to side-effect in any of the switch cases. To help the reader, I'd like to scope these calculations into an inline closure, like:

```swift
let sectionTitle: String = {
    switch pluginStatus {
    case .active:
        return Localization.activeSectionTitle
    case .inactive:
        return Localization.inactiveSectionTitle
    case .networkActive:
        return Localization.networkActiveSectionTitle
    case .unknown:
        return "" // This case should not happen
   }
}()
```

What do you think about this approach?

Reply:
I think this is pretty neat 👍 Will certainly adopt this approach in my future work!